### ISSUE TITLE
Wait throws IOError: closed stream

### ISSUE THREAD

**Issue by osteenbergen** [spriteCloud/lapis-lazuli#28]
Seems to happen when setting `:context`

``` ruby
browser.wait(
  :like => [:a,:class,"disabled"],
  :context => some_div
  :condition => :while
)
```

```
IOError: closed stream
ruby-2.2.1/gems/lapis_lazuli-0.8.2/lib/lapis_lazuli/browser/wait.rb:76:in `internal_encoding'
ruby-2.2.1/gems/lapis_lazuli-0.8.2/lib/lapis_lazuli/browser/wait.rb:76:in `dump'
ruby-2.2.1/gems/lapis_lazuli-0.8.2/lib/lapis_lazuli/browser/wait.rb:76:in `block (2 levels) in internal_wait'
ruby-2.2.1/gems/lapis_lazuli-0.8.2/lib/lapis_lazuli/browser/wait.rb:74:in `times'
ruby-2.2.1/gems/lapis_lazuli-0.8.2/lib/lapis_lazuli/browser/wait.rb:74:in `block in internal_wait'
ruby-2.2.1/gems/watir-webdriver-0.8.0/lib/watir-webdriver/wait.rb:40:in `block in until'
ruby-2.2.1/gems/watir-webdriver-0.8.0/lib/watir-webdriver/wait.rb:75:in `call'
ruby-2.2.1/gems/watir-webdriver-0.8.0/lib/watir-webdriver/wait.rb:75:in `block in run_with_timer'
ruby-2.2.1/gems/watir-webdriver-0.8.0/lib/watir-webdriver/wait/timer.rb:14:in `wait'
ruby-2.2.1/gems/watir-webdriver-0.8.0/lib/watir-webdriver/wait.rb:74:in `run_with_timer'
ruby-2.2.1/gems/watir-webdriver-0.8.0/lib/watir-webdriver/wait.rb:40:in `until'
ruby-2.2.1/gems/lapis_lazuli-0.8.2/lib/lapis_lazuli/browser/wait.rb:104:in `internal_wait'
ruby-2.2.1/gems/lapis_lazuli-0.8.2/lib/lapis_lazuli/browser/wait.rb:41:in `wait'
```

**Comment by sjieg** [comment]
Same as #35

**Comment by sjieg** [comment]
Update on research:
The error comes from the `Marshal.dump` function used in /browser/wait.rb:49.

The dump is failing because of the context element in `options[:selectors][0][:context]`
```
#<Watir::Div:0x..fe01fc718 located=true selector={:class=>"tsf-p", :tag_name=>"div", :index=>0}>
```

Next looking into why it is not failing for `find` with a context.

**Comment by sjieg** [comment]
The find function is not using Marshal.dump anywhere. So I've removed the code
```
opts = Marshal.load(Marshal.dump(options))
```
And directly put the options into the send function
```
results = send(find_func.to_sym, options)
```
It seems to work. Going to write some test cases in the self tester to make sure this didn't break any of the funtctions.

**Comment by jfinkhaeuser** [comment]
```ruby
opts = Marshal.load(Marshal.dump(options))
```
is shorthand for a deep copy of the options Hash. If you pass the options directly, the find functions may modify the Hash. So calling find twice with the same Hash may mean that the second time there are fields in the Hash you don't want.

TL;DR it's not safe.

**Comment by jfinkhaeuser** [comment]
It's perfectly possible that `Marshal` isn't suitable for dealing with context objects. Maybe the hackaround would be to do:

```ruby
context = options.delete(:context)
opts = Marshal.load(Marshal.dump(options))
opts[:context] = options[:context] = context
```

**Comment by sjieg** [comment]
Good stuff, updated the code accordingly:
```
# Removing context from the options to preven an error in Marshal.dump
context = options[:selectors][0][:context]
options[:selectors][0].delete(:context)
# [...]
opts = Marshal.load(Marshal.dump(options))
# Putting :context back in the selector if it was used
opts[:selectors][0][:context] = context unless context.nil?
results = send(find_func.to_sym, opts)
```

Also created some new scenario's in the self test where :context in combination with wait() is tested. This brought up a new issue, which I'll create separately.
